sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,492 @@
+      bytes_to_request = min(self._remaining(), self.buffer_size)
+      self.buffer_start_position = self.position
+      try:
+        result = self._get_segment(self.position, bytes_to_request)
+      except Exception as e:  # pylint: disable=broad-except
+        tb = traceback.format_exc()
+        logging.error(
+            ('Exception while fetching %d bytes from position %d of %s: '
+             '%s\n%s'),
+            bytes_to_request, self.position, self.path, e, tb)
+        raise
+
+      self.buffer = result
+      return
+
+  def _remaining(self):
+    return self.downloader.size - self.position
+
+  def close(self):
+    """Close the current file."""
+    self.closed = True
+    self.download_stream = None
+    self.downloader = None
+    self.buffer = None
+
+  def _get_segment(self, start, size):
+    """Get the given segment of the current file."""
+    if size == 0:
+      return ''
+    end = start + size - 1
+    self.downloader.get_range(start, end)
+    value = self.download_stream.getvalue()
+    # Clear the cStringIO object after we've read its contents.
+    self.download_stream.truncate(0)
+    assert len(value) == size
+    return value
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, exception_type, exception_value, traceback):
+    self.close()
+
+  def seek(self, offset, whence=os.SEEK_SET):
+    """Set the file's current offset.
+
+    Note if the new offset is out of bound, it is adjusted to either 0 or EOF.
+
+    Args:
+      offset: seek offset as number.
+      whence: seek mode. Supported modes are os.SEEK_SET (absolute seek),
+        os.SEEK_CUR (seek relative to the current position), and os.SEEK_END
+        (seek relative to the end, offset should be negative).
+
+    Raises:
+      IOError: When this buffer is closed.
+      ValueError: When whence is invalid.
+    """
+    self._check_open()
+
+    self.buffer = ''
+    self.buffer_start_position = -1
+
+    if whence == os.SEEK_SET:
+      self.position = offset
+    elif whence == os.SEEK_CUR:
+      self.position += offset
+    elif whence == os.SEEK_END:
+      self.position = self.downloader.size + offset
+    else:
+      raise ValueError('Whence mode %r is invalid.' % whence)
+
+    self.position = min(self.position, self.downloader.size)
+    self.position = max(self.position, 0)
+
+  def tell(self):
+    """Tell the file's current offset.
+
+    Returns:
+      current offset in reading this file.
+
+    Raises:
+      IOError: When this buffer is closed.
+    """
+    self._check_open()
+    return self.position
+
+  def _check_open(self):
+    if self.closed:
+      raise IOError('Buffer is closed.')
+
+  def seekable(self):
+    return True
+
+  def readable(self):
+    return True
+
+  def writable(self):
+    return False
+
+
+# TODO: Consider using cStringIO instead of buffers and data_lists when reading
+# and writing.
+class BufferedWriter(object):
+  """A class for writing files from stateless services.
```

Comment:
<!--new_thread; commit:970afe0523d6d5673489a70e801be6553e0ffc33; resolved:0-->
"writing files to"?

(actually, what does "stateless" here mean?)

Reply:
Stateless refers to the file API: there is no long-lived file handle, the filesystem server doesn't store our current file position, etc.